rts/motoko-rts/src/stabilization/layout.rs

```diff
@@ -0,0 +1,375 @@
+//!
+//! Pointers are serialized as skewed offsets in that space.
+//! Scalar and pointer values are serialized as 64-bit `StableValue`
+//! for a long-term perspective, even if main memory operates on 32-bit.
+//! A 32-bit program stores `StableValue` that fit into 32-bit, and
+//! scalar values beyond this width are boxed.
+//! This means that a 32-bit and 64-bit program can always upgrade from
+//! a 32-bit version, while the downgrade from 64-bit to a 32-bit program
+//! is not supported and detected.
+//!
+//! Each object uses a `StableTag` as header an is followed by
+//! the object payload as outlined in the corresponding Rust structs.
+//! Some objects, such as `StableArray`, `StableObject`, `StableBlob`,
+//! and `StableBigNum`
+//!
+//! Not all heap memory object types are stabilized because some
+//! of them are not stable types. New object types can be added
+//! with backwards compatibility but encoding changes to existing stable
+//! data types must be handled extra care to ensure backwards compatibility.
+
+use crate::types::{
+    block_size, size_of, Tag, Value, Words, TAG_ARRAY, TAG_BIGINT, TAG_BITS32, TAG_BITS64,
+    TAG_BLOB, TAG_CONCAT, TAG_MUTBOX, TAG_OBJECT, TAG_OBJ_IND, TAG_REGION, TAG_VARIANT, TRUE_VALUE,
+};
+
+use self::{
+    stable_array::StableArray, stable_bigint::StableBigInt, stable_bits32::StableBits32,
+    stable_bits64::StableBits64, stable_blob::StableBlob, stable_concat::StableConcat,
+    stable_mutbox::StableMutBox, stable_obj_ind::StableObjInd, stable_object::StableObject,
+    stable_region::StableRegion, stable_variant::StableVariant,
+};
+
+use super::{
+    reader_writer::{ScanStream, StableMemorySpace, WriteStream},
+    StableMemoryAccess,
+};
+
+mod stable_array;
+mod stable_bigint;
+mod stable_bits32;
+mod stable_bits64;
+mod stable_blob;
+mod stable_concat;
+mod stable_mutbox;
+mod stable_obj_ind;
+mod stable_object;
+mod stable_region;
+mod stable_variant;
+
+/// Object tag in stable format. Encoded as a `i32` value.
+#[repr(C)]
+#[derive(Clone, Copy)]
+enum StableTag {
+    _None = 0,
+    Array = 1,
+    MutBox = 2,
+    Object = 3,
+    Blob = 4,
+    Bits32 = 5, // Note: Can be removed in 64-bit heap support.
+    Bits64 = 6,
+    Region = 7,
+    Variant = 8,
+    Concat = 9,
+    BigInt = 10,
+    ObjInd = 11,
+}
+
+const _: () = assert!(core::mem::size_of::<StableTag>() == core::mem::size_of::<i32>());
+
+impl StableTag {
+    fn deserialize(tag: Tag) -> StableTag {
+        match tag {
+            TAG_ARRAY => StableTag::Array,
+            TAG_MUTBOX => StableTag::MutBox,
+            TAG_OBJECT => StableTag::Object,
+            TAG_BLOB => StableTag::Blob,
+            TAG_BITS32 => StableTag::Bits32,
+            TAG_BITS64 => StableTag::Bits64,
+            TAG_REGION => StableTag::Region,
+            TAG_VARIANT => StableTag::Variant,
+            TAG_CONCAT => StableTag::Concat,
+            TAG_BIGINT => StableTag::BigInt,
+            TAG_OBJ_IND => StableTag::ObjInd,
```

Comment:
I'm not sure if this is used to stabilize from main memory into stable memory, but it its, then I thing it might be missing a case for TAG_CLOSURE.
Consider:

```
stable let x : {} = { f = func () {} );
```
The type of x is an empty record, so stable, but its value is a non-empty record containing a closure. It the graph copy has to either drop the field, or translate it a (fake) closure or other dummy value. My hope it that a fake closure or dummy value is fine because it will be statically unreachable anyway (the static type doesn't reveal the field and we don't allow down-casts).

Reply:
Thanks a lot for noticing this. It is related to your previous question, maybe more such types are affected. Probably, we could simply omit fields with non-stable types, as those are unreachable by the stable type definitions.

Reply:
Great!

Re dropping fields, the static compatibility  check actually does not let you drop stable variables, to avoid resurrecting them later at an incompatible type (IIRC). The most you can do is promote them to type Any. Perhaps your type table based compatibility check missed that aspect.  But you are probably right  and the issue remains that the data will now never be removed. With Candid, once promoted to Any, the data would essentially be erased on destabilisation. With graph copy, it's retained. Unless we can somehow arrange that stable fields of type Any are not considered as roots by the GC.

Reply:
Oh, this is an interesting point: Currently, the static compatibility check allows dropping fields in records (meaning declaring less fields than the original definition). I thought this would be safe because it would be a sub-type and thus compatible from memory perspective. Of course, there could be sharing, and other references may still point to the object by its full type. Deleting a field would only be conservatively possible, we can be sure that such a field is not declared in any (compatible) record type in that program, e.g. by collecting the field hashes used in a program version and checking whether the deserialized or resurrected object contains such old fields. But maybe a miss a point here?

Reply:
**Minor improvement**: The implementation now drops the content of `Any`-typed actor fields.